File: app/src/main/java/pl/laskowski/marcin/model/dividers/AntidiagonalDivider.java
package pl.laskowski.marcin.model.dividers;

import pl.laskowski.marcin.model.Field;
import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.model.Sudoku;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

/**
 * Created by Marcin Laskowski.
 */

public class AntidiagonalDivider implements SudokuDivider {

    @Override
    public Set<Region> divideIntoRegions(Sudoku sudoku) {
        int limit = Math.min(sudoku.sizeX(), sudoku.sizeY());
        List<Field> fields = new ArrayList<>();
        for (int i = 0; i < limit; i++) {
            fields.add(sudoku.at(limit - i - 1, i));
        }
        Region region = new Region(fields);
        return Collections.singleton(region);
    }

}

File: app/src/main/java/pl/laskowski/marcin/model/dividers/AntidiagonalDivider.kt
package pl.laskowski.marcin.model.dividers

import pl.laskowski.marcin.model.Board
import pl.laskowski.marcin.model.Field
import pl.laskowski.marcin.model.Region
import java.util.ArrayList
import kotlin.math.min

class AntidiagonalDivider : SudokuDivider {

    override fun divideIntoRegions(board: Board): Set<Region> {
        val limit = min(board.sizeX(), board.sizeY())
        val fields: MutableList<Field> = ArrayList()
        for (i in 0 until limit) {
            fields.add(board.at(limit - i - 1, i)!!)
        }
        val region = Region(fields)
        return setOf(region)
    }
}